File: android-app/app/src/main/java/com/braininput/capture/data/PendingCaptureStore.kt
package com.braininput.capture.data

import android.content.Context
import com.braininput.capture.util.CaptureLog
import java.io.File

class PendingCaptureStore(context: Context) {

    private val pendingDir = File(context.filesDir, "pending_captures").apply { mkdirs() }

    fun save(payload: CapturePayload) {
        try {
            val file = File(pendingDir, "${payload.idempotencyKey}.json")
            file.writeText(payload.toJson())
            CaptureLog.store("Saved pending capture: ${payload.idempotencyKey}")
        } catch (e: Exception) {
            CaptureLog.store("Failed to save pending: ${e.message}")
        }
    }

    fun loadAll(): List<CapturePayload> {
        return try {
            pendingDir.listFiles { file -> file.extension == "json" }
                ?.mapNotNull { file ->
                    try {
                        CapturePayload.fromJson(file.readText())
                    } catch (e: Exception) {
                        CaptureLog.store("Failed to parse ${file.name}: ${e.message}")
                        null
                    }
                } ?: emptyList()
        } catch (e: Exception) {
            CaptureLog.store("Failed to load pending captures: ${e.message}")
            emptyList()
        }
    }

    fun remove(idempotencyKey: String) {
        File(pendingDir, "$idempotencyKey.json").delete()
    }

    val pendingCount: Int
        get() = pendingDir.listFiles { file -> file.extension == "json" }?.size ?: 0
}

File: macos-app/Sources/ThoughtInput/Core/PendingCaptureStore.swift
import Foundation

final class PendingCaptureStore: Sendable {
    private let directory: URL

    init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first ?? FileManager.default.temporaryDirectory
        directory = appSupport.appendingPathComponent("ThoughtInput/pending", isDirectory: true)
        do {
            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        } catch {
            CaptureLog.store.error("Failed to create pending captures directory: \(error.localizedDescription)")
        }
    }

    @discardableResult
    func save(_ capture: PendingCapture) -> Bool {
        let file = directory.appendingPathComponent("\(capture.payload.idempotencyKey).json")
        do {
            let data = try JSONEncoder().encode(capture)
            try data.write(to: file, options: .atomic)
            CaptureLog.store.info("Saved pending capture: \(capture.payload.idempotencyKey)")
            return true
        } catch {
            CaptureLog.store.error("Failed to save pending capture: \(error.localizedDescription)")
            return false
        }
    }

    func loadAll() -> [PendingCapture] {
        do {
            let files = try FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
                .filter { $0.pathExtension == "json" }
            return files.compactMap { file in
                guard let data = try? Data(contentsOf: file) else { return nil }
                // Try new format first, fall back to legacy bare CapturePayload
                if let pending = try? JSONDecoder().decode(PendingCapture.self, from: data) {
                    return pending
                }
                // Legacy format: wrap bare payload with a nil destination (will use active destination on retry)
                if let payload = try? JSONDecoder().decode(CapturePayload.self, from: data) {
                    return PendingCapture(payload: payload, destinationID: UUID(), destinationSnapshot: Destination(
                        name: "Legacy",
                        isActive: true,
                        type: .restNoAuth(RESTNoAuthConfig(endpointURL: ""))
                    ))
                }
                return nil
            }
        } catch {
            CaptureLog.store.error("Failed to load pending captures: \(error.localizedDescription)")
            return []
        }
    }

    func remove(idempotencyKey: String) {
        let file = directory.appendingPathComponent("\(idempotencyKey).json")
        try? FileManager.default.removeItem(at: file)
    }

    var pendingCount: Int {
        (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" }.count) ?? 0
    }
}
